Compiler crash (missing body) with declaration of inline proc

Traceback:

	/home/user/nimskull/compiler/nim.nim(154) nim
	/home/user/nimskull/compiler/nim.nim(89) handleCmdLine
	/home/user/nimskull/compiler/front/main.nim(539) mainCommand
	/home/user/nimskull/compiler/front/main.nim(491) compileToBackend
	/home/user/nimskull/compiler/front/main.nim(217) commandCompileToC
	/home/user/nimskull/compiler/backend/cbackend.nim(406) generateCode
	/home/user/nimskull/compiler/backend/cbackend.nim(478) generateCode
	/home/user/nimskull/compiler/backend/cbackend.nim(289) emit
	/home/user/nimskull/lib/system/assertions.nim(38) failedAssertImpl
	/home/user/nimskull/lib/system/assertions.nim(28) raiseAssert
	/home/user/nimskull/lib/system/fatal.nim(50) sysFatal
	Error: unhandled exception: /home/user/nimskull/compiler/backend/cbackend.nim(289, 10) `prc.body.code != nil` missing body [AssertionDefect]

I can only trigger this bug with 4 separate files. Minimum example is attached (try to compile `test.nim`).
[missing_body.tar.gz](https://github.com/user-attachments/files/16097122/missing_body.tar.gz)

I think the key lines are these:

	proc foo*[T](utype: typedesc[T]): var seq[T] {.inline.}
	import ./bar
	proc foo*[T](utype: typedesc[T]): var seq[T] {.inline.} = discard

Notice how the proc is declared before importing `bar` because it's part of a cyclic module import and both modules are meant to be able to use each other procs.

A workaround is easy. The biggest problem with this bug is the lack of any feedback on what the problem might be. Disallowing this kind of use of inline functions is fine, but it should have an appropriate error message.

without `{.inline.}` it produces a linker error:
```
...........................................................
~/Programming/skulls/missing_body/foo.nim(2, 8) Warning: imported and not used: 'bar' [UnusedImport]
~/Programming/skulls/missing_body/all.nim(1, 8) Warning: imported and not used: 'foo' [UnusedImport]
/usr/bin/ld: ~/.cache/nimskull/test_d/@mbar.nim.c.o: in function `crash__bar_2':
@mbar.nim.c:(.text+0x67): undefined reference to `foo__bar_3'
/usr/bin/ld: ~/Programming/skulls/missing_body/test: hidden symbol `foo__bar_3' isn't defined
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
execution of an external program 'gcc   -o ~/Programming/skulls/missing_body/test  ~/.cache/nimskull/test_d/stdlib_ansi_c.nim.c.o ~/.cache/nimskull/test_d/stdlib_memory.nim.c.o ~/.cache/nimskull/test_d/stdlib_digitsutils.nim.c.o ~/.cache/nimskull/test_d/stdlib_assertions.nim.c.o ~/.cache/nimskull/test_d/stdlib_dollars.nim.c.o ~/.cache/nimskull/test_d/stdlib_repr_v2.nim.c.o ~/.cache/nimskull/test_d/stdlib_io.nim.c.o ~/.cache/nimskull/test_d/stdlib_system.nim.c.o ~/.cache/nimskull/test_d/@mbar.nim.c.o ~/.cache/nimskull/test_d/@mtest.nim.c.o    -ldl' failed with exit code '1'
```

--- a/compiler/ast/report_enums.nim
+++ b/compiler/ast/report_enums.nim
@@ -360,6 +360,7 @@ type
 
     rsemCannotInstantiate
     rsemCannotInstantiateWithParameter
+    rsemCannotInstantiateForwarded
     rsemCannotGenerateGenericDestructor
     rsemUndeclaredField
     rsemExpectedOrdinal

--- a/compiler/front/cli_reporter.nim
+++ b/compiler/front/cli_reporter.nim
@@ -790,6 +790,10 @@ proc reportBody*(conf: ConfigRef, r: SemReport): string =
           r.ownerSym.name.s
         )
 
+    of rsemCannotInstantiateForwarded:
+      result = "cannot instantiate generic procedure forward-declared in " &
+               "another module"
+
     of rsemTypeKindMismatch:
       result = r.str
       result.add  " got '$1'" % typeToString(r.actualType)

--- a/compiler/sem/seminst.nim
+++ b/compiler/sem/seminst.nim
@@ -401,6 +401,11 @@ proc generateInstance(c: PContext, fn: PSym, pt: TIdTable,
   n[genericParamsPos] = c.graph.emptyNode
   var oldPrc = genericCacheGet(c.graph, fn, entry[], c.compilesContextId)
   if oldPrc == nil:
+    if sfForward in fn.flags and fn.itemId.module != c.module.itemId.module:
+      localReport(c.config, info, reportSem(rsemCannotInstantiateForwarded))
+      # don't abort instantiation; let it complete for the sake of error
+      # correction (check/suggest)
+
     # we MUST not add potentially wrong instantiations to the caching mechanism.
     # This means recursive instantiations behave differently when in
     # a ``compiles`` context but this is the lesser evil. See

--- a/tests/lang_callable/generics/mgeneric_cycle_forward.nim
+++ b/tests/lang_callable/generics/mgeneric_cycle_forward.nim
@@ -1,0 +1,5 @@
+import tgeneric_cycle_forward
+
+forwarded[int]() # was already instantiated; works
+# try to instantiate the incomplete generic routine:
+forwarded[float]()

--- a/tests/lang_callable/generics/tgeneric_cycle_forward.nim
+++ b/tests/lang_callable/generics/tgeneric_cycle_forward.nim
@@ -1,0 +1,20 @@
+discard """
+  description: '''
+    Ensure instantiating foreign, incomplete generic procedures leads to a
+    proper error.
+  '''
+  errormsg: "cannot instantiate generic procedure forward-declared in another module"
+  file: "mgeneric_cycle_forward.nim"
+  line: 5
+"""
+
+proc forwarded*[T]()
+
+# instantiate with `int` before starting the import cycle
+forwarded[int]()
+
+import mgeneric_cycle_forward # start the recursive import
+
+# complete the forward declaration:
+proc forwarded[T]() =
+  discard

